### ISSUE TITLE
Passing maps of attributes to components

### ISSUE THREAD

**Issue by lnr0626** [msaraiva/surface#95]
I'm trying to add some accessibility tags to the links generated by `LivePatch`/`LiveRedirect`, but they don't current expose those attributes as properties. I also don't think it makes sense to add them as properties as that'll quickly get unmanageable. 

I was trying to find a way similar to the `Link` opts property, so it would look something like:
```
<LivePatch to="/url" class="some-class" opts={{ aria-label="This link does some things" }}>
    <icon></icon>
</LivePatch>
```
and the component would just render those opts as key=value attributes. However, I couldn't figure out how to make that work. Is this possible?

**Comment by msaraiva** [comment]
Hi @lnr0626!

Do you know if phoenix's `live_patch/2` already accepts blocks containing live components? As far as I can remember, that was the reason for not using it. Maybe this has already been fixed in Phoenix 1.5. Otherwise, we could implement `<LivePatch>` using `~L` instead of `~H` and generate the attributes dynamically.

For a general solution, maybe we could add an `:attributes` directive similar to the proposed `:props` that inject attributes for regular HTML nodes. WDYT?

**Comment by lnr0626** [comment]
That would be ideal, but I don't think `live_patch/2` accepts blocks containing live components currently - [their response](https://github.com/phoenixframework/phoenix_live_view/issues/693) indicated that it's not feasible given the way LV components work.

That sounds like a reasonable direction, though one thought that comes to mind is could the `:props` functionality be extended to support this use case? i.e. both are similar capabilities - one for `Component`s, one for standard html elements. Looking through the comments on #2, it looks like someone asked something similar (https://github.com/msaraiva/surface/issues/2#issuecomment-614304684).

Taking another look, it appears that #68 might actually already accomplish this.

**Comment by msaraiva** [comment]
> one thought that comes to mind is could the `:props` functionality be extended to support this use case? i.e. both are similar capabilities - one for `Component`s, one for standard html elements.

We could extend `:props` but I think it would be better to have a separate directive `:attrs` or `:attributes` with different semantics. 

The reason is that a `property` is not the same as an `attribute`. A property is just a piece of information that is passed to a component. The component then decides what to do with that information. It can use it to add an attribute to the root node or to any other child node. It can add multiple attributes or maybe none! There's no imposed relation between a component's property and any HTML attribute of any element generated by the component.

However, if we define a separate `:attrs` directive, we can then have that relation explicitly, i.e. the values passed to `:attrs` will automatically become attributes in the HTML node.

I think it's important to keep the distinction between components' properties and HTML attributes so the user can always expect a predictable behaviour.

Does it make sense?

**Comment by lnr0626** [comment]
That does - I think I have a better idea what you're suggesting as well. I thought you were suggesting an `:attributes` directive that would be used on plain HTML elements in a similar fashion to the `:props` directive; but you were suggesting an `:attributes` directive that would instead add those values to the root element that the component generates. i.e.:

```elixir
~H"""
<LivePatch label="Link" to="/" :attributes={{ %{ "aria-label" => "This link does some things" }} />
"""
```
would generate:
```html
<a href="/" aria-label="This link does some things">
    Link
</a>
```

I agree that we wouldn't want to extend the `:props` functionality to support something like that for the reasons you've enumerated.

I was thinking instead that this functionality could be accomplished once the `:props` proposal is in by adding an additional property to the component (say `opts` or similar), and then the component would use `:props` internally to accomplish this. As an example:

```elixir
defmodule Surface.Components.LivePatch do
    use Surface.Component

    property(class, :to, required: true)
    # Other properties removed to simplify
    property(opts, :map, default: %{})

    def render(assigns) do
        ~H"""
        <a :props={{ @opts }} href={{ @to }}>Link to somewhere</a>
        """
    end
end
```

I suppose the benefit of the `:attributes` directive would be that components don't need to participate/allow this for it to work. I have the feeling that there'll be some edge cases with `:attributes` - the first couple that come to mind are handling multiple root elements (I believe that's allowed for stateless components), and components that just call a function to generate their html (e.g. `Link`).

What do you think about re-evaluating this once we have the `:props` directive in?

**Comment by msaraiva** [comment]
Yeah, sounds good to me. 👍

**Comment by lnr0626** [comment]
circling back on this, I think it may be resolved with #117, and #120 adds support for the specific use case I had in mind. unless you can think of a use case not covered by the `:attrs` directive, i'll go ahead and close this issue once #120 is merged

**Comment by msaraiva** [comment]
Yeah, feel free to close this one.
